nixos/modules/security/acme/default.nix

```diff
@@ -142,43 +120,59 @@ let
           chown -R ${user}:${data.group} "$fixpath"
         fi
       done
-    '') certConfigs));
-  in {
-    description = "Fix owner and group of all ACME certificates";
-
-    serviceConfig = commonServiceConfig // {
-      # We don't want this to run every time a renewal happens
-      RemainAfterExit = true;
-
-      # StateDirectory entries are a cleaner, service-level mechanism
-      # for dealing with persistent service data
-      StateDirectory = [ "acme" "acme/.lego" "acme/.lego/accounts" ];
-      StateDirectoryMode = 755;
-      WorkingDirectory = "/var/lib/acme";
-
-      # Run the start script as root
-      ExecStart = "+" + (pkgs.writeShellScript "acme-fixperms" script);
-    };
-  };
-  lockfilePrepareService = {
-    description = "Manage lock files for acme services";
-
-    # ensure all required lock files exist, but none more
-    script = ''
-      GLOBIGNORE="${lib.concatStringsSep ":" concurrencyLockfiles}"
-      rm -f -- *
-      unset GLOBIGNORE
-
-      xargs touch <<< "${toString concurrencyLockfiles}"
-    '';
-
-    serviceConfig = commonServiceConfig // {
-      # We don't want this to run every time a renewal happens
-      RemainAfterExit = true;
-      WorkingDirectory = lockdir;
-    };
-  };
-
+    '') certConfigs))
+  );
+
+  # This is defined with lib.mkMerge so that we can separate the config per function.
+  setupService = lib.mkMerge [
+    {
+      description = "Set up the ACME certificate renewal infrastructure";
+      script = lib.mkBefore ''
+        ${lib.optionalString cfg.defaults.enableDebugLogs "set -x"}
+        set -euo pipefail
+      '';
+      serviceConfig = commonServiceConfig // {
+        # This script runs with elevated privileges, denoted by the +
+        # ExecStartPre is used instead of ExecStart so that the `script` continues to work.
+        ExecStartPre = "+${lib.getExe privilegedSetupScript}";
+
+        # We don't want this to run every time a renewal happens
+        RemainAfterExit = true;
+
+        # StateDirectory entries are a cleaner, service-level mechanism
+        # for dealing with persistent service data
+        StateDirectory = [ "acme" "acme/.lego" "acme/.lego/accounts" ];
+        StateDirectoryMode = "0755";
+
+        # Creates ${lockdir}. Earlier RemainAfterExit=true means
+        # it does not get deleted immediately.
+        RuntimeDirectory = "acme";
+        RuntimeDirectoryMode = "0700";
+
+        # Generally, we don't write anything that should be group accessible.
+        # Group varies for most ACME units, and setup files are only used
+        # under the acme user.
+        UMask = "0077";
+      };
+    }
+
+    # In order to avoid race conditions creating the CA for selfsigned certs,
+    # we have a separate service which will create the necessary files.
+    (lib.mkIf cfg.preliminarySelfsigned {
+      path = [ pkgs.minica ];
+      # Working directory will be /tmp
+      script = ''
+        test -e ca/key.pem || minica \
```

Comment:
Also inline minica path:
```nix
      # Working directory will be /tmp
      script = ''
        test -e ca/key.pem || ${lib.getExe pkgs.minica} \
``` 

(Can't put a native suggestion for some reason)

Reply:
Same as [the other thread](https://github.com/NixOS/nixpkgs/pull/355087#discussion_r1845169317), fine by me.